src/Microsoft.OData.Edm/Csdl/Serialization/EdmModelCsdlSchemaJsonWriter.cs

```diff
@@ -0,0 +1,1292 @@
+            this.jsonWriter.WriteOptionalProperty("$OpenType", entityType.IsOpen, CsdlConstants.Default_OpenType);
+
+            // It MAY contain the members $HasStream, The value of $HasStream is one of the Boolean literals true or false. Absence of the member means false
+            this.jsonWriter.WriteOptionalProperty("$HasStream", entityType.HasStream, CsdlConstants.Default_HasStream);
+        }
+
+        /// <summary>
+        /// Write Complex Type Object
+        /// </summary>
+        /// <param name="complexType">The Edm complex type.</param>
+        internal override void WriteComplexTypeElementHeader(IEdmComplexType complexType)
+        {
+            // A complex type is represented as a member of the schema object whose name is the unqualified name of the complex type
+            this.jsonWriter.WritePropertyName(complexType.Name);
+
+            // whose value is an object
+            this.jsonWriter.WriteStartObject();
+
+            // The complex type object MUST contain the member $Kind with a string value of ComplexType.
+            this.jsonWriter.WriteRequiredProperty("$Kind", CsdlConstants.Element_ComplexType);
+
+            // It MAY contain the members $BaseType
+            this.jsonWriter.WriteOptionalProperty("$BaseType", complexType.BaseComplexType(), this.TypeDefinitionAsJson);
+
+            // It MAY contain the members $Abstract, The value of $Abstract is one of the Boolean literals true or false. Absence of the member means false.
+            this.jsonWriter.WriteOptionalProperty("$Abstract", complexType.IsAbstract, CsdlConstants.Default_Abstract);
+
+            // It MAY contain the members $OpenType, The value of $OpenType is one of the Boolean literals true or false. Absence of the member means false.
+            this.jsonWriter.WriteOptionalProperty("$OpenType", complexType.IsOpen, CsdlConstants.Default_OpenType);
+        }
+
+        /// <summary>
+        /// Write $Key
+        /// </summary>
+        internal override void WriteDeclaredKeyPropertiesElementHeader()
+        {
+            // The value of $Key is an array with one item per key property.
+            this.jsonWriter.WritePropertyName("$Key");
+
+            // Its value is an array.
+            this.jsonWriter.WriteStartArray();
+        }
+
+        /// <summary>
+        /// Write the Key property
+        /// </summary>
+        /// <param name="property"></param>
+        internal override void WritePropertyRefElement(IEdmStructuralProperty property)
+        {
+            // Key properties without a key alias are represented as strings containing the property name.
+            // Key properties with a key alias are represented as objects with one member whose name is the key alias and whose value is a string containing the path to the property.
+            // TODO: It seems the second one is not supported.

```

Comment:
>// TODO: It seems the second one is not supported. [](start = 12, length = 50)

do we have an issue to track?

Reply:
an issue created at [https://github.com/OData/odata.net/issues/1817](https://github.com/OData/odata.net/issues/1817)

---
In reply to: [443713186](https://github.com/OData/odata.net/pull/1775#discussion_r443713186) [](ancestors = 443713186)